### ISSUE TITLE
Compiling vAmiga on Windows

### ISSUE THREAD

**Issue by dirkwhoffmann** [dirkwhoffmann/vAmiga#611]
As part of my efforts to improve general code quality, I would like to make vAmiga's code base compatible with Visual C++. Since I don't use Windows regularly, I am unsure about the best approach. 

The current situation is as follows: I develop vAmiga on a Mac using Xcode. In addition to the Xcode project file, there are several Makefiles that can be used to compile vAmiga on Linux. There is already a GitHub Action setup that automatically compiles the code in a Ubuntu VM when code is checked in on the main branch.

My first thought was to setup a similar GitHub Action script which uses a Windows VM as target environment. However, I am unsure what I need to do here. I know that Visual C++ comes with nmake. Is it feasible and does it make sense to setup a Windows build environment the same way as I've setup the Ubuntu environment? Or is there another standard approach I am not aware of? 

Any input from Window users are welcome.

**Comment by emoon** [comment]
I would recommend using a GH action that uses the default windows image as it has the msvc compiler.

I don't use cmake myself, but it's a quite popular option these days to target many platforms. With cmake you can generate vs compatible files that can be used with compilation.

**Comment by mras0** [comment]
+1 to the recommendation to use CMake. I use that myself for my projects. It generates project files for Visual Studio (and other IDEs) and makes it easy to use another compiler (like Clang on windows for example). Its major downside is the terrible scripting language and somewhat arcane syntax (though it's gotten a bit better over the years), but this shouldn't be an issue for this project, since you're not doing anything complicated / out of the ordinary. On the plus side it's widely used and you can easily find answers to most questions you might have.

I've already tried building with MSVC (using a simple cmakelists file), and it doesn't look too bad, but there are some issues:

- Use of GCC specific features (not many), but basically anything that starts with double underscore (e.g. __builtin_expect). The builtins are easy to work around, but the use of `__attribute__` for alignment won't work (perhaps you can use the standard alignof/alignas as appropriate here).
- Relying on POSIX stuff (e.g. `apra/inet.h`, `dirent.h`, `unistd.h` and `ssize_t`). All of these either have direct equivalents, comparability libraries or even standard replacements (say `std::filesystem`).
- Having the file `IO.h` (from the `Utilities` directory) in the include path shouldn't be an issue, but it is. On windows (with MSVC and clang) it ends up shadowing a necessary file (that defines some `unistd.h` replacements, e.g. `write`). This shouldn't be an issue, but it is :)

I also saw, what's probably a MSVC bug (since Clang and GCC accepts it), but I haven't tried debugging it:
```
1>C:\Prog\amiga\Misc\vAmiga\Emulator\CPU\Moira\MoiraExec_cpp.h(139,1): error C2668: 'moira::Moira::writeM': ambiguous call to overloaded function
1>C:\Prog\amiga\Misc\vAmiga\Emulator\CPU\Moira\MoiraDataflow_cpp.h(318,8): message : could be 'void moira::Moira::writeM<moira::MODE_AN,moira::Word,0>(moira::u32,moira::u32)'
1>C:\Prog\amiga\Misc\vAmiga\Emulator\CPU\Moira\MoiraDataflow_cpp.h(278,8): message : or       'void moira::Moira::writeM<moira::MEM_DATA,moira::Word,0>(moira::u32,moira::u32)'
1>C:\Prog\amiga\Misc\vAmiga\Emulator\CPU\Moira\MoiraExec_cpp.h(139,1): message : while trying to match the argument list '(moira::u32, moira::u32)'
1>C:\Prog\amiga\Misc\vAmiga\Emulator\CPU\Moira\MoiraInit_cpp.h(187): message : see reference to function template instantiation 'void moira::Moira::execAddRgEa<moira::ADD,moira::MODE_AN,moira::Word>(moira::u16)' being compiled
```

Anyway, none of these issues are insurmountable, and I'll be happy to help out if you have any questions.

**Comment by mras0** [comment]
With a bit of hacks (and disabling minor features), I've managed to get latest vAmiga (33e24693aed2ffea21943a1044f59aad815dd353) running on Windows using MinGW (specifically [this port](https://nuwen.net/mingw.html) version)

![image](https://user-images.githubusercontent.com/9952017/142926046-cf0102ca-cb37-4db4-9d10-06c743c07d45.png)

There were a bit more issues that the ones I mentioned in my last post, but again, nothing too bad. Frontend/driver is just a straight forward SDL2 implementation. I think to get it working with MSVC a good first stepping stone is ironing out issues that occur with MinGW, but you decide the course obviously :)

**Comment by emoon** [comment]
Personally I think it would be good to have it compiling in msvc also. The combo of clang, gcc and msvc usually irons out quite a bunch of general C/C++ "issues" and makes the codebase more portable and compatible.

**Comment by mras0** [comment]
@emoon: 100% agree, but some work (and changes to the emulator core)  is required to make that happen.

Add in running cleanly with `-fsanitize=address/undefined` (I've tried the first one, and basic stuff seemed fine). 

Getting to 0 warnings with `-Wextra` and `-pedantic` (not to mentioned `/W4` with MSVC) will also require extra effort (that may not be worth it, selectively disabling warning is probably appropriate for now).

**Comment by emoon** [comment]
`-pedantic` is not worth it. It contains many "compiler internal" warnings that is usually not worth fixing `-Wall`, `-Wextra` and msvc `/W4` is totally worth it in my book at least :)

I also ran into this the other day. https://lgtm.com I haven't tested it myself, but it's free for open source projects and I will give it a go for sure.

**Comment by dirkwhoffmann** [comment]
I agree. The goal is a code base that 

- compiles with `clang`, `gcc`, and `msvc`.
- issues no warning with `-Wall`, `-Wextra`, `/W4`.
- runs cleanly with `-fsanitize=address/thread/undefined`.

The lgtm project looks very interesting. I haven't heard about it yet.

**Comment by dirkwhoffmann** [comment]
I've set up an experimental cmake branch: https://github.com/dirkwhoffmann/vAMIGA/tree/cmake

As a starting point, I've added two `CMakeLists.txt` files, one in `Emulator` and the other one in `Emulator/Utilities`. There is also a new Dummy.cpp file to make the (cropped) project compile. 

The top-level file looks as follows:

```CMake
cmake_minimum_required(VERSION 3.22 FATAL_ERROR)

# Set project name
project(vAmiga)

# Add include paths
include_directories(${CMAKE_SOURCE_DIR}/..)

# Specify the C++ standard
set(CMAKE_CXX_STANDARD 20)
set(CMAKE_CXX_STANDARD_REQUIRED True)

# List all source files
set(files Dummy.cpp)
add_executable(vAmiga ${files})

# Add sub directories
add_subdirectory(Utilities)

target_link_libraries(vAmiga Utilities)
```

The other one looks looks like so: 

```CMake
set(files
  Chrono.cpp
  Concurrency.cpp
  Exception.cpp
  SSEUtils.cpp
  MemUtils.cpp
  Checksum.cpp
  IO.cpp
  Parser.cpp
)

add_library(Utilities ${files})
```

When running `cmake` followed by `make`, I get: 

```
hoff@Dirks-MBP-2 vAmigaBuild % make clean; make        
[  9%] Building CXX object Utilities/CMakeFiles/Utilities.dir/Chrono.cpp.o
[ 18%] Building CXX object Utilities/CMakeFiles/Utilities.dir/Concurrency.cpp.o
[ 27%] Building CXX object Utilities/CMakeFiles/Utilities.dir/Exception.cpp.o
[ 36%] Building CXX object Utilities/CMakeFiles/Utilities.dir/SSEUtils.cpp.o
[ 45%] Building CXX object Utilities/CMakeFiles/Utilities.dir/MemUtils.cpp.o
[ 54%] Building CXX object Utilities/CMakeFiles/Utilities.dir/Checksum.cpp.o
[ 63%] Building CXX object Utilities/CMakeFiles/Utilities.dir/IO.cpp.o
[ 72%] Building CXX object Utilities/CMakeFiles/Utilities.dir/Parser.cpp.o
[ 81%] Linking CXX static library libUtilities.a
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: libUtilities.a(Exception.cpp.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: libUtilities.a(Exception.cpp.o) has no symbols
[ 81%] Built target Utilities
[ 90%] Building CXX object CMakeFiles/vAmiga.dir/Dummy.cpp.o
[100%] Linking CXX executable vAmiga
[100%] Built target vAmiga
```

Am I on the right track? My current setup will produce a lib file (.a) for all sub directories. Is this the recommended way to go?

**Comment by mras0** [comment]
I won't die defending `-pedantic` :), I usually just enable it for my own projects (because it's easier if you enable stuff like that early on, like with all warnings). It doesn't matter for code quality only potential portability issues.

Hadn't heard about lgtm before, thanks for the pointer. It was super easy to get working, took me literally 2 minutes and zero effort besides from clicking a few buttons. On the plus side it seems to not give any false positives and isn't "noisy", on the downside it didn't find anything interesting. Because it's apparently possible to enable for any github project, I also tried for vAmiga, but it couldn't deduce how to build it.

I think MSVC's `/analyze` (that I forgot to mention) finds a lot more useful stuff, the downside is that it's (way) more noisy.

[Coverity](https://scan.coverity.com/) also offers free scanning of open source projects, and it's (I conjecture) way more advanced that lgtm, but it's harder to setup and get working (at least locally). If the code isn't prepared for it, you also have to dig through many false positives, but I've seen it find very real errors that had been in code bases for a long time and slipped through code review.

**Comment by mras0** [comment]
> Am I on the right track? My current setup will produce a lib file (.a) for all sub directories. Is this the recommended way to go?

Yes, that's totally fine, and probably the most "proper" way to do it. However, if you just want to get something working quickly, you can also just create one big library containing the emulator. The advantage of splitting is up is that you get a better grip of dependencies between the different libraries (modules) and potentially faster compile times when you make changes. The downside is you have to manage dependencies to get it working.

You can see what I've used here:

[vAmigaSDL.zip](https://github.com/dirkwhoffmann/vAmiga/files/7590088/vAmigaSDL.zip)

Note: Using builtin CMake features (like you're doing) is preferred over hacking compiler flags directly, I just do that out of old (bad) habbits. Also I'm building from a sibling directory because I don't want to pollute my "vAmiga" dir.

**Comment by emoon** [comment]
> I won't die defending -pedantic :), I usually just enable it for my own projects (because it's easier if you enable stuff like that early on, like with all warnings). It doesn't matter for code quality only potential portability issues.

So the problem I ran into when doing this is that when you have other people trying to use your code. If they don't have exact matching compiler version it may report different errors. And in cases when you disable some warnings it won't work on compiler x.y because it doesn't know about it. 

What I usually end up doing is try with `-pedantic` and then enable a bunch of extra warnings that isn't included with `extra. all` as middle ground.

**Comment by dirkwhoffmann** [comment]
> You can see what I've used here: vAmigaSDL.zip

Great! This helps a lot.

 I've also seen that I can use the `target_sources` command which makes my CMake files even simpler. I'll try to setup all CMakeLists that way and compile the project locally. Once this is done, I'll add proper GitHub Actions for performing automatic builds.

**Comment by dirkwhoffmann** [comment]
Update: 
- All Makefiles have been replaced by CMakeLists.txt files.
- There are two GitHub actions now: One for Ubuntu and one for Windows.

As expected, the Windows build fails.

**Comment by mras0** [comment]
I've narrowed down the only issue (the one I mentioned [above](https://github.com/dirkwhoffmann/vAmiga/issues/611#issuecomment-973103658)) that can't be fixed by simple `#ifdef`-ing code out/providing replacements.

It boils down to MSVC not accepting the following (see on [compiler explorer](https://godbolt.org/z/qvMxWen8v)):
```
enum M { ma };
enum N { na };
struct Foo {
    template<M m> static void f() {}
    template<N n> static void f() {}
};
int main() {
    Foo::f<ma>();
    Foo::f<na>();
}
```

As I wrote, it's likely a MSVC bug, but it's been too long since I've done any C++ language lawyering to say for certain. I couldn't find any bug reports or similar though.

**Comment by emoon** [comment]
I asked about the above and someone posted this which was a bit interesting https://godbolt.org/z/ohd9GnoYo now one can see that MSVC is "loosing" the enum type along the way (and only uses the underlying `int` type) 

When asking about how to solve this, the consensus seems to be "Don write code like this".

**Comment by dirkwhoffmann** [comment]
I am basically stuck at the moment, because I cannot install Visual Studio without errors in my VM. Running all this Microsoft typical "Click here to fix me, run this to repair me" stuff seems to make everything worse. I will trash my existing VM and reinstall Windows from scratch.

**Comment by emoon** [comment]
Should work fine from a VM. I have done it a few times in the past.

**Comment by mras0** [comment]
Here's a crude patch (for a2f681fe533e345319b43803597a94b24038a803) that at least gets things compiling: [crude_patch.zip](https://github.com/dirkwhoffmann/vAmiga/files/7604960/crude_patch.zip). Don't actually include it, but it should highlight where there are issues, and some way to work around some of them.

I can get this far (kick 1.3, no disk inserted) before crashing with those:
![image](https://user-images.githubusercontent.com/9952017/143493521-9c5f7574-d1ce-4803-a52c-f57587afe9f8.png)

You'll need some kind of executable to actually debug problems. Probably at least a headless version of vAmiga (so you can use retroshell scripts to get e.g. the boot image to see the MSVC version is still working), but a simple SDL gui will probably also help.

I probably broke something with my changes, though it still seems to work under Linux, so take it FWIW.

**Comment by dirkwhoffmann** [comment]
I've wiped out my VM and reinstalled Windows. This time, I was able to install Visual Studio without errors. However,  cmake complains about a non-matching generator. I'm unsure what is meant by "used previously". cmake was running for the very first time on a fresh Windows installation. Does anybody know how to deal with this? 

```
1> [CMake] CMake Error: Error: generator : Ninja
1> [CMake] Does not match the generator used previously: Visual Studio 17 2022
1> [CMake] Either remove the CMakeCache.txt file and CMakeFiles directory or choose a different binary directory.
1> Fehler bei der Ausführung von "C:\WINDOWS\system32\cmd.exe /c "%SYSTEMROOT%\System32\chcp.com 
```

**Comment by mras0** [comment]
It seems weird that it wants to use the "ninja" generator.

You should be able to select something like "x64 native tools command prompt for VS 2019" from the start menu (press whatever maps to the windows key and type "2019" and it should show up).

Make a directory somewhere convenient and cd into it, the run `cmake \path\to\vAmiga` should now do something sane (latest visual studio ships with its own CMake now, but only a v3.20 variant, so you may need to change your CMakeLists.txt file like in my patch).

If some of the above fails you maybe didn't get all the right components installed. In that case launch "visual studio installer" and make sure "Desktop development with C++" is enabled".

**Comment by dirkwhoffmann** [comment]
> x64 native tools command prompt for VS 2019

This did the trick. Previously I was running CMake from the developer terminal that can be opened from within Visual Studio. 

```
C:\Users\Dirk Hoffmann\source\repos\vAmiga\EmuBuild>cmake ../Emulator
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.18363.
-- The C compiler identification is MSVC 19.30.30705.0
-- The CXX compiler identification is MSVC 19.30.30705.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.30.30705/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.30.30705/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/Dirk Hoffmann/source/repos/vAmiga/EmuBuild
```

**Comment by mras0** [comment]
Great! I just noticed you're using a newer version than me (2022 vs 2019), so there might have been some changes (using the developer prompt should have worked). Maybe they even fixed the enum issue. I'll give that a shot tomorrow. Either way the project/solution you got should hopefully work.

**Comment by dirkwhoffmann** [comment]
I've startet to analyse the MSVC error output. The first error is related to `ssize_t` which doesn't exist on Windows. I've mapped my custom integer type `isize` to `ssize_t`, because I wanted `isize` to be a "performance type" of the target machine.  

@mras0: In your build, you've mapped `ssize_t`to `ptrdiff_t` which I think is a good workaround. In fact, using `ssize_t` might not be a good idea at all as the standard only assures that `ssize_t` is capable of storing the negative number -1. We don't have the guarantee that it can store -2, although I doubt that there is any architecture out there with this limitation. 

What I want to achieve with `isize` is exactly what `int_fast32_t` is doing (it provides a type with at least 32 bit and the additional property that no other type with at least 32 bit can be processed faster). Unfortunately, mapping `isize` to `int_fast32_t` has drawbacks. E.g., on macOS, this type is mapped to `int` (4 bytes). On compiler explorer, I got it mapped to `long` (8 bytes). This means that I would get into trouble with all my `printf` statements. With the current mapping (to `ssize_t`), I can use `%zd` as placeholder, but this would no longer work with `int_fast32_t`.

**Comment by mras0** [comment]
The solution to the printing problem is easy, but ugly. You would have to use the "Format macro constants" from [cinttypes](https://en.cppreference.com/w/cpp/types/integer). Extra ugly in C++ because of the note: "Because C++ interprets a character immediately following a string literal as a user-defined string literal,"

**Comment by dirkwhoffmann** [comment]
> easy, but ugly.

    std::printf("%+" PRIdFAST32 "\n", 42);

Seriously? No way that I integrate such code 😂.

A nice solution would be to use the C++20 format library instead of `printf`. 
https://www.youtube.com/watch?v=KeS1ehp9IiI

Unfortunately, it seems like most compilers still don't support it.

Maybe it's best to not overcomplicate things here and simply map `isize` to `long` and `usize` to `unsigned long`. This should give us the desired result on all architectures that are relevant for this project.

**Comment by emoon** [comment]
Personally I always do this in my projects

```
#include <stdint.h>

typedef uint32_t u32;
typedef uint64_t u64;
typedef int8_t i8;
typedef float f32;
typedef double f64;
...
```

Etc. Just make everything nice, short and to the point imo.

**Comment by dirkwhoffmann** [comment]
> Just make everything nice, short and to the point imo.

Yes, I agree. In my opinion, simplicity is the number one rule in software engineering. However, there is a reason why I defined `isize` with varying width. On macOS, I want this type to map to 64 bit, because of Swift. In this case, `isize` directly maps to the standard Swift integer type (Swift is so strict about types that it's sometimes a nightmare). But on architectures such as the Raspberry, I want it to be 32 bit because of performance reasons. The same holds for WASM which is still ILP32 (I think).

**Comment by emoon** [comment]
Yeah, I know the arguments for `isize` as iterator and such because of platform size difference. The thing is that I have seen very little it terms of better performance due to it, so I prefer to have the exact size as it shows the intention better to the reader imo.

**Comment by mras0** [comment]
> > easy, but ugly.
> 
> ```
> std::printf("%+" PRIdFAST32 "\n", 42);
> ```
> 
> Seriously? No way that I integrate such code 😂.

Yeah, this is one place where iostreams (with some helper stuff for formatting) really win out. 

> 
> A nice solution would be to use the C++20 format library instead of `printf`. https://www.youtube.com/watch?v=KeS1ehp9IiI
> 
> Unfortunately, it seems like most compilers still don't support it.

You could consider using https://github.com/fmtlib/fmt I it was the testbed for the standard library.

> 
> Maybe it's best to not overcomplicate things here and simply map `isize` to `long` and `usize` to `unsigned long`. This should give us the desired result on all architectures that are relevant for this project.

Just be aware that long is always 32-bits on windows (even 64-bit windows), so they're usually the wrong choice IMHO. If I'm sticking to standard types I use "int" when I want 32-bits and "long long" if I want 64-bits (yes, I know this is not guaranteed, but that can be checked with a static_assert).

But like emoon I usually use the standard types with exact sizes and "size_t/ptrdiff_t" for dealing with containers.

**Comment by emoon** [comment]
> Yeah, this is one place where iostreams (with some helper stuff for formatting) really win out. 

iostreams *really* bloats the code tho and makes the code (imo) much harder to read.

**Comment by dirkwhoffmann** [comment]
Now, MSVC issues a `C4201: nonstandard extension used : nameless struct/union` for TOD.h:

```
typedef union
{
    struct
    {
        u8 lo;
        u8 mid;
        u8 hi;
    };
    u32 value;
}
Counter24;
```

Aren't nameless structs / unions part of the standard since C++ 11?

**Comment by emoon** [comment]
Seems this is a bug in msvc https://developercommunity.visualstudio.com/t/warning-level-4-generates-a-bogus-warning-c4201-no/103064 for now I would suggest you just disable the warning (add `/wd4201` to the compiler commandline)

**Comment by mras0** [comment]
That bug report is for C11 (which does allow anonymous structs). GCC also complains if you use `-pedantic`:

>  warning: ISO C++ prohibits anonymous structs [-Wpedantic]

as does clang:

>warning: anonymous structs are a GNU extension [-Wgnu-anonymous-struct]

Anonymous unions are allowed though.

**Comment by dirkwhoffmann** [comment]
> but if you want to be completely correct, you're probably better off creating a normal struct with lo/mid/high and synthesizing the value on demand with a member function

I use anonymous structs quite a lot. What I like about them is the possibility to directly access an array via multiple APIs. In VirtualC64, I use this approach to access disk data. Data can be read from the disk data array directly by either specifying a track number or a halftrack number. This is so nice (at least in my view) that I want to keep it. Hence, I think I just disable the warning. 

The next item to work on is to replace the POSIX API (dirent.h) by the C++ filesystem library. I didn't use the filesystem library back then, because Apple has implemented it pretty late. I think it wasn't available until Xcode 11 came out.

**Comment by dirkwhoffmann** [comment]
The dependency to `dirent.h` is gone. As a drawback, the deployment target had to be pushed from macOS 10.14 to macOS 10.15 which means that vAmiga will no longer run on macOS Mojave. I think we can live with that.

**Comment by mras0** [comment]
> I use anonymous structs quite a lot. What I like about them is the possibility to directly access an array via multiple APIs. In VirtualC64, I use this approach to access disk data. Data can be read from the disk data array directly by either specifying a track number or a halftrack number. This is so nice (at least in my view) that I want to keep it. Hence, I think I just disable the warning.

Totally reasonable, and I don't think now would be the time to make such a change anyway. 


BTW just got things more or less working now. The issue was in `AmigaFile::init`:

`std::ifstream stream(path);`

needs to be:

`std::ifstream stream(path, std::ifstream::binary);`

I can see the same issue in a few other places where binary files are opened. The flag actually makes a difference on Windows, as otherwise `"\r\n"` gets translated to `"\n"` (and the other way round when writing). If you use `fopen` with binary files you want to add a `"b"` for the same effect.

More or less pedantic notes:

In [Macros.h](https://github.com/dirkwhoffmann/vAmiga/blob/fe5f29abf37d5692f9cefb0e9c33e10118c8d23d/Emulator/Utilities/Macros.h#L19) you should probably use `#ifdef _MSC_VER` instead of `#ifdef _WIN32`, since `__assume` is a MSVC-only thing. 

In [`loadFile`](https://github.com/dirkwhoffmann/vAmiga/blob/fe5f29abf37d5692f9cefb0e9c33e10118c8d23d/Emulator/Utilities/IOUtils.cpp#L222) you allocate memory using `new[]` but then in e.g. [`FsDevice::importDirectory`](https://github.com/dirkwhoffmann/vAmiga/blob/fe5f29abf37d5692f9cefb0e9c33e10118c8d23d/Emulator/FileSystems/FSDevice.cpp#L881) you use plain `delete`. That technically has to be `delete[] buffer` though it'll probably work. In general I'd avoid using raw allocated memory like that and use either `std::vector` or at the very least `std::unique_ptr<u8[]>`.

**Comment by dirkwhoffmann** [comment]
> needs to be: std::ifstream stream(path, std::ifstream::binary);

Great!! I've added the flag to all calls to this function (except for one where a script file is read).

I've also fixed the `delete []` bug and the `_MSC_VER` macro issue.

**Comment by dirkwhoffmann** [comment]
The next error is related to pipes. They are utilised by the screen recorder to communicate with `ffmpeg`:

     unlink(videoPipePath().c_str());

The problem here is more general, because the screen recorder is very Unix centric. E.g., it relies on `ffmpeg` being installed in `/usr/local/bin`. I guess it's best to take the screen recorder out in Windows builds, at least for now.

**Comment by mras0** [comment]
> The next error is related to pipes. They are utilised by the screen recorder to communicate with `ffmpeg`:
> 
> ```
>  unlink(videoPipePath().c_str());
> ```
> 
> The problem here is more general, because the screen recorder is very Unix centric. E.g., it relies on `ffmpeg` being installed in `/usr/local/bin`. I guess it's best to take the screen recorder out in Windows builds, at least for now.

Yeah, definitely postpone that part for now. There are possibly ways to make the code work without too many changes. Probably same deal with the "fmemopen" stuff used for DMS files. Both are IMHO nice to have, but shouldn't hold you up.

I've been testing a few demos, and things mostly seems to work as well as they do under linux via WSL2. I was even about to consider reporting a bug for TBL/Eon corrupting the display when you don't insert disk 2 in time, but that seems consistent with latest WinUAE (@emoon: Get on it ;) )

![image](https://user-images.githubusercontent.com/9952017/143620548-73cb2c3f-1013-47d6-a248-da704aa8c859.png)

**Comment by mras0** [comment]
Seems like you're making good progress!

A quick hack for the use of "fmemopen" with DMS files is using this "xdms/pfile.c" instead:

[pfile.zip](https://github.com/dirkwhoffmann/vAmiga/files/7614033/pfile.zip)

With the following changes in `DMSFile.cpp`:

```
extern "C" {
unsigned short extractDMS(const u8* in, u32 inSize, u8** out, u32* outSize, int verbose);
}
void
DMSFile::finalizeRead()
{
    u8* out;
    u32 outSize;
    if (!extractDMS(data, size, &out, &outSize, DMS_DEBUG))
        adf = new ADFFile(out, outSize);
    else
        adf = nullptr;
    free(out);
    
    if (!adf) throw VAError(ERROR_UNKNOWN);
}
```

Note: it seems like you're missing a destructor for `DMSFile` (Not related to this change) that deletes `adf`.

I also noticed:

 - `FSBlock::exportFileHeaderBlock`: `fopen` should be `"wb"` (lost in your recent update)
 - `Emulator/Paula/Audio/Muxer.h`: `fraction` should be initialized to `0`
 - `Recorder::recordVideo` seems leaks  to `data`

**Comment by dirkwhoffmann** [comment]
> With the following changes in DMSFile.cpp:

Cool! So we can have DMS support in the Windows build.

Right now, the compiler stops here: 

```C++
  // Trigger Blitter interrupt if this is the termination cycle
  if ((instr & BLTDONE) && !birq) {
      paula.scheduleIrqRel(INT_BLIT, DMA_CYCLES(1));
  }
```

MSVC issues a C4127: Conditional Expression is Constant

In theory, I like this warning, because it reminds me to replace statements such as `if (instr & BLTDONE)` by `if constexpr (instr & BLTDONE)`. However, I cannot do this in the above statement for obvious reasons. It seems like MSVC wants me to write code like this:

```C++
  // Trigger Blitter interrupt if this is the termination cycle
  if constexpr (instr & BLTDONE) {
      if ( && !birq) {
          paula.scheduleIrqRel(INT_BLIT, DMA_CYCLES(1));
      }
  }
```

Personally, I don't see a reason why constant and non-constant parts shouldn't be combined, so I think I have to disable this warning in the end.

**Comment by dirkwhoffmann** [comment]
> `Recorder::recordVideo` seems leaks to `data`

Oups, this needs to be cleaned up thoroughly. Memory is allocated dynamically in every frame (and never released). How embarrassing 🙈.

**Comment by mras0** [comment]
> MSVC issues a C4127: Conditional Expression is Constant
> 
> In theory, I like this warning, because it reminds me to replace statements such as `if (instr & BLTDONE)` by `if constexpr (instr & BLTDONE)`. However, I cannot do this in the above statement for obvious reasons.

Yeah, and it can also help catch something like `unsigned_num < 0` or `char_val > 127`, and sometimes less obvious things. However as you've discovered it can be very noisy with lots of false positives (and I usually end up disabling it as well).

C4244 is the same way in most codebases.

One thing we did on a project I previously worked on was to have a special build configuration that enabled most warnings (and `/ANALYZE`) with a whitelist of "allowed" deviations. That way you can keep the normal build warning-free and still occasionally check the noisy one, audit the warnings in batches and either fix them or add them to the whitelist.

**Comment by dirkwhoffmann** [comment]
Milestone reached: Both the GCC build and the MSVC build pass:

<img width="604" alt="Bildschirmfoto 2021-11-30 um 13 41 00" src="https://user-images.githubusercontent.com/12561945/144049901-b3cb6d9b-622c-4a7e-9113-c98feea6e48a.png">

The Windows build enables `/W4 /WX` with the following warnings disabled:

- `C4100`: Unreferenced formal parameter 
- `C4201`: Nonstandard extension used : nameless struct/union
- `C4324` structure was padded due to __declspec(align())
- `C4458`: Declaration of 'xxx' hides class member.

Furthermore, I've disabled C4127 for file `Moira.cpp`:
- `C4127`: Conditional Expression is Constant

For all other files, I kept `C4127` enabled. 

Does it make sense to setup an additional GitHub action with MinGW?

**Comment by mras0** [comment]
Great effort 👍, things seem to work (only very light testing done though) with my hacky SDL frontend with MSVC 2022 (v. 17.0.1). If I try with GCC 9.3.0 on WSL it crashes on startup unless I build the vAmiga library myself, but I'm investigating that.

A few notes:

- It would be good if you could change the use of `CMAKE_SOURCE_DIR` to `CMAKE_CURRENT_SOURCE_DIR` in `Emulator/CMakeLists.txt`. That way other projects can use `add_subdirectory` to include it.
- Maybe consider renaming the library to `vAmigaCore` (or something like that) as most frontends would probably like to call their executable `vAmiga` and that conflicts with the library
- Users of the library also currently have to replicate the include path, so you should consider using `target_include_directories(vAmiga PUBLIC ....)` instead of `include_directories`
- You may also want to consider whether you always want to compile things (on non-MSVC platforms) with `-O3`, that seems to be something CMAKE_BUILD_TYPE (or other flags) should determine instead.
- Minor point: You currently require CMake 3.21, but I don't think you're using features that weren't available in e.g. 3.16. Just unnecessary friction.
- I've created a gist with an example on how the recorder could be borough to work on Windows as well: https://gist.github.com/mras0/a2b91f0315ec48d72c89c6c98b132f8b (shouldn't need to many changes, but I don't know how/if you would like to incorporate them)


> Does it make sense to setup an additional GitHub action with MinGW?

It would allow people allergic to MSVC to compile vAmiga under windows (if these people exists), and it doesn't require much more effort to get working. It also exercises the code path where `WIN32` is defined but `_MSC_VER` isn't, which will find other portability issues. E.g. recorder compile there and `arpa/inet.h` doesn't exist.

For `ntohs` etc. maybe consider using e.g. `__builtin_bswap16` (of course this needs to only be done on little endian architectures, but there are probably more issues there once you look for them. For Windows this shouldn't be an issue since I think the only possible big endian "windows" machine is Xbox360).

Either way, with very little effort I just got things running with MinGW on the current version, and that seems to worth holding on to.

**Comment by dirkwhoffmann** [comment]
> For ntohs etc. maybe consider using

It's fascinating to see that there is still no common way to carry out such an elementary task on all common platforms. I think it's best if I stop using `ntohs` etc. and replace it by something like this: 

```
#ifdef _MSC_VER
#define SWAP16 _byteswap_ushort
#define SWAP32 _byteswap_ulong
#define SWAP64 _byteswap_uint64
#else
#define SWAP16  __builtin_bswap16
#define SWAP32  __builtin_bswap32
#define SWAP64  __builtin_bswap64
#endif

template<typename T> T bigEndian(T x);

template<> 
uint16_t bigEndian(uint16_t x) 
 {
    if constexpr (std::endian::native == std::endian::big) {
        return x;
    } else {
        return SWAP16(x);
    }
}

template<> 
uint32_t bigEndian(uint32_t x) 
 {
    if constexpr (std::endian::native == std::endian::big) {
        return x;
    } else {
        return SWAP32(x);
    }
}

template<> 
uint64_t bigEndian(uint64_t x) 
{
    if constexpr (std::endian::native == std::endian::big) {
        return x;
    } else {
        return SWAP64(x);
    }
}
```

**Comment by mras0** [comment]
I guess from C++23 we'll get `std::byteswap`. BTW ntohs etc. are also available on Windows ([MSDN](https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-ntohs)), but they require including `winsock2.h` (which pulls in a lot of stuff including `windows.h`). This really pollutes the global namespace and is IMHO best avoided. If the functions weren't time critical, I'd recommend just putting them in a separate cpp file.

**Comment by dirkwhoffmann** [comment]
> If the functions weren't time critical

Unfortunately, they are. They are used whenever the virtual CPU or Agnus access memory. This is the main reason why I want to stick to the build-in macros. I want to make sure that the compiler translates it into the `swap` machine instruction.

I know that compilers are smart today. E.g., if the following code is compiled in clang, 

```
int foo(uint32_t x)
{
    return ((x >> 24) & 0xFF) | ((x >> 8) & 0xFF00) | ((x << 8) & 0xFF0000) | ((x << 24) & 0xFF000000);
}
```

the compiler detects that the code reverses the byte order and translates it into a swap instruction:

```
foo(unsigned int):                                # @foo(unsigned int)
        mov     eax, edi
        bswap   eax
        ret
```

What I like about the build-in macros (whatever they are called on different architecture) is that it is almost certain that the compiler creates the `swap` instruction. If I go for the generic code, I have to trust that my compiler is smart enough to recognise the meaning of the code.

**Comment by emoon** [comment]
> I've been testing a few demos, and things mostly seems to work as well as they do under linux via WSL2. I was even about to consider reporting a bug for TBL/Eon corrupting the display when you don't insert disk 2 in time, but that seems consistent with latest WinUAE (@emoon: Get on it ;) )

This has actually been fixed in the final (non yet released) version, still waiting for the finished graphics.. x years and counting :(

Also re endian swap https://godbolt.org/z/dzbz5T6hG all big compilers (gcc, msvc, clang) figures this out.

**Comment by dirkwhoffmann** [comment]
> - It would be good if you could change the use of `CMAKE_SOURCE_DIR` to `CMAKE_CURRENT_SOURCE_DIR` in `Emulator/CMakeLists.txt`. That way other projects can use `add_subdirectory` to include it.

done

> - Maybe consider renaming the library to `vAmigaCore` (or something like that) as most frontends would probably like to call their executable `vAmiga` and that conflicts with the library

done

> - Users of the library also currently have to replicate the include path, so you should consider using `target_include_directories(vAmiga PUBLIC ....)` instead of `include_directories`

done 

> - You may also want to consider whether you always want to compile things (on non-MSVC platforms) with `-O3`, that seems to be something CMAKE_BUILD_TYPE (or other flags) should determine instead.

done

> - Minor point: You currently require CMake 3.21, but I don't think you're using features that weren't available in e.g. 3.16. Just unnecessary friction.

done (I run into this issue myself in my Linux Mint VM)

> - I've created a gist with an example on how the recorder could be borough to work on Windows as well: https://gist.github.com/mras0/a2b91f0315ec48d72c89c6c98b132f8b (shouldn't need to many changes, but I don't know how/if you would like to incorporate them)

I'll open a separate issue for this, so it won't be forgotten...

**Comment by dirkwhoffmann** [comment]
Automatic MinGW builds are up and running:

<img width="906" alt="Bildschirmfoto 2021-12-01 um 16 00 07" src="https://user-images.githubusercontent.com/12561945/144258238-f93efe59-9f88-4c5f-8cff-3c4a4dd8fd8b.png">

**Comment by mras0** [comment]
Awesome! It's really easy to integrate now, so if anyone wants to step up to the plate and create a cross-platform GUI this would no longer be the excuse :)

Only very final small change is needed: In the master [CMakeLists.txt](https://github.com/dirkwhoffmann/vAmiga/blob/master/Emulator/CMakeLists.txt#L36) file the definition of `SCREEN_RECORDER` needs to be `PUBLIC` since it affects the size the "Amiga" object (that's why it crashed for me, but only under Linux).

I've checked MSVC 2022 (32-bit and 64-bit, v. 19.30.30705), Linux (under WSL2) with gcc 9.3.0, MinGW (64-bit, v. 9.3.0) and Clang 13.0.0 on Windows (using the MSVC standard library). The last one required moving `FSDeviceDescriptor::FSDeviceDescriptor()` to the cpp file (it complained about constructing `partitions` with unknown size), but otherwise no issues.

**Comment by dirkwhoffmann** [comment]
I close this for now, because there are no Windows related issues at the moment. Feel free to reopen anytime.

**Comment by dirkwhoffmann** [comment]
Recently, all MinGW builds started to fail with a very strange error:

```
D:/a/_temp/msys64/mingw64/include/c++/12.1.0/bits/basic_string.h:3541:36,
[96](https://github.com/dirkwhoffmann/vAmiga/runs/6642674061?check_suite_focus=true#step:4:97)
    inlined from 'std::string util::makeUniquePath(const std::string&)' at D:/a/vAmiga/vAmiga/Emulator/Utilities/IOUtils.cpp:105:43:
[97](https://github.com/dirkwhoffmann/vAmiga/runs/6642674061?check_suite_focus=true#step:4:98)
D:/a/_temp/msys64/mingw64/include/c++/12.1.0/bits/char_traits.h:431:56: error: 'void* __builtin_memcpy(void*, const void*, long long unsigned int)' accessing 9223372036854775810 or more bytes at offsets [2, 9223372036854775807] and 1 may overlap up to 9223372036854775813 bytes at offset -3 [-Werror=restrict]
[98](https://github.com/dirkwhoffmann/vAmiga/runs/6642674061?check_suite_focus=true#step:4:99)
  431 |         return static_cast<char_type*>(__builtin_memcpy(__s1, __s2, __n));
```

I can't really see how this error maps to my code. Also, I don't remember that I've made any changes to this particular code lately. Does somebody have an idea?

Line 105 is this line

    auto suffix = "." + extractSuffix(path);

in the following code:

```
string
extractSuffix(const string &s)
{
    auto idx = s.rfind('.');
    auto pos = idx != string::npos ? idx + 1 : 0;
    auto len = string::npos;
    return s.substr(pos, len);
}

string
makeUniquePath(const string &path)
{
    auto prefix = stripSuffix(path);
    auto suffix = "." + extractSuffix(path);
    
    string index = "";
    for (isize nr = 2; util::fileExists(prefix + index + suffix); nr++) {
        index = std::to_string(nr);
    }

    return prefix + index + suffix;
}

```

**Comment by mras0** [comment]
I think it's a GCC regression, look related to [this bug report](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=105545).

**Comment by dirkwhoffmann** [comment]
> I think it's a GCC regression, look related to [this bug report](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=105545).

Thanks! I've "fixed" it by passing `-Wno-restrict` as compile option. 

However, this was just the tip of the iceberg. The next error is just as weird (I think): 

```
D:/a/vAmiga/vAmiga/Emulator/Zorro/DiagBoard.cpp:239:28: error: no matching function for call to 'find(std::vector<unsigned int>::iterator, std::vector<unsigned int>::iterator, u32&)'
[111](https://github.com/dirkwhoffmann/vAmiga/runs/6643653261?check_suite_focus=true#step:4:112)
  239 |         auto it = std::find(tasks.begin(), tasks.end(), ptr1);
```

To me it seems like the MinGW image utilises a screwed up gcc version.

**Comment by mras0** [comment]
The two errors look like `<algorithm>` isn't included. It was probably included transitively by another standard header previously, and now they've "tightened it up" a bit.

**Comment by dirkwhoffmann** [comment]
> The two errors look like `<algorithm>` isn't included

Indeed. Adding a couple of `#include <algorithm>` made the build pass 😎.

**Comment by dirkwhoffmann** [comment]
With my latest CPU changes I get a strange MinGW error: 

```
In file included from D:/a/vAmiga/vAmiga/Emulator/CPU/Moira/Moira.cpp:26:
D:/a/vAmiga/vAmiga/Emulator/CPU/Moira/MoiraExec_cpp.h: In member function 'void moira::Moira::execMove7(moira::u16) [with moira::Core C = moira::M68020; moira::Instr I = moira::MOVE; moira::Mode M = moira::MODE_DI; moira::Size S = moira::Long]':
D:/a/vAmiga/vAmiga/Emulator/CPU/Moira/MoiraExec_cpp.h:2621:23: error: dangling pointer to 'data' may be used [-Werror=dangling-pointer=]
 2621 |     reg.sr.n = NBIT<S>(u64(data));
      |                ~~~~~~~^~~~~~~~~~~
D:/a/vAmiga/vAmiga/Emulator/CPU/Moira/MoiraExec_cpp.h:2614:13: note: 'data' declared here
 2614 |     u32 ea, data;
      |             ^~~~
```

Here is the relevant code: 

 ```
    u32 ea, data;

    int src = _____________xxx(opcode);
    int dst = ____xxx_________(opcode);

    if (!readOp<C, M, S, STD_AE_FRAME>(src, &ea, &data)) return;

    reg.sr.n = NBIT<S>(u64(data));
    reg.sr.z = ZERO<S>(u64(data));
```

I have no idea how a dangling pointer may be used here. I am using no pointer at all in line 2621.

Maybe somebody has an idea?

**Comment by mras0** [comment]
Seems bogus. I don't get that warning with this version: g++-12 (Ubuntu 12.1.0-2ubuntu1~22.04) 12.1.0. After a painful installation process I got the msys64 version of mingw64 installed (which seems to be the version github is using), but I don't get that warning instead I get: `Moira.cpp:284:27: error: array subscript [0, 7] is outside array bounds of 'const moira::u32 [8]' {aka 'const unsigned int [8]'}` which seems equally bogus. Ignoring that warning it compiles without further warnings/errors (version reported: g++ (Rev3, Built by MSYS2 project) 12.1.0).

If I were you, I'd switch to an earlier g++ version while this is sorted out upstream if it's configurable in github actions. Alternatively disable the warning on that file for this case, i.e. something like `if (MINGW AND CMAKE_CXX_COMPILER_VERSION EQUALS 12.1.0) ...` (don't remember the exact CMake syntax).

P.S. Next time it would be nice if the issue wasn't in a file that took literal minutes to compile :)

**Comment by dirkwhoffmann** [comment]
> Seems bogus.

Thanks for having a look at it and confirming that it was not due to some stupid coding error I was too blind to see.

After adding

    target_compile_options(vAmigaCore PUBLIC -Wno-dangling-pointer)

to the MinGW settings, the build passes. 

> Next time it would be nice if the issue wasn't in a file that took literal minutes to compile :)

Yes, my CPU is pushing template usage to the limits. However, it seems to pay off as it makes Moira run even faster than Musashi. I am a little bit concerned about the growing memory footprint. Because the CPU revision is a template parameter, the number of synthesizes instruction handlers grows proportionally with the number of supported CPU models. Fortunately, all modern compilers seem to check for identical functions bodies and eliminate duplicates.

**Comment by mras0** [comment]
The long build time is just a fact of template heavy C++ code, not a criticism. Just annoying that it always seems to happen in the translation units that take the longest to compile :)

It seems that `-Wdangling-pointer` is new GCC 12 so I figured it would be worth a bug report, but since it's not reproducible for me I won't bother. (BTW [creduce](https://embed.cs.utah.edu/creduce/) is very useful when doing compiler bug reports - most easily done on the pre-processed source code).

**Comment by mras0** [comment]
Speaking of compiler (linker) issues. I can no longer link an executable with latest vAmiga version using MSVC (19.33.31629). The library compiles just fine, but when linking it fails due to:
```
private: unsigned int __cdecl moira::Moira::dasmRead<0>(unsigned int &)" (??$dasmRead@$0A@@Moira@moira@@AEAAIAEAI@Z) referenced in function "private: struct moira::Ea<11,0> __cdecl moira::Moira::Op<11,0>(unsigned short,unsigned int &)" (??$Op@$0L@$0A@@Moira@moira@@AEAA?AU?$Ea@$0L@$0A@@1@GAEAI@Z)
```
Which is `Moira::dasmRead<Unsized>`. I assume this is a MSVC bug and a work around is to add
```
#ifdef _MSC_VER
template <>
u32 Moira::dasmRead<Unsized>(u32& addr)
{
    assert(false);
    return 0;
}
#endif
```
to the top of MoiraDasm_cpp.h (using `fatalError` gives too many unreachable code warnings).

To catch something like this it might be an idea to add a very simple test target that just instantiates an "Amiga" object. If a few automatic tests could also be run (using a non-copyrighted test ROM) that might be an extra bonus. Think that might have been discussed earlier. I may cook up a pull request if you think something like that would be a good idea.

**Comment by dirkwhoffmann** [comment]
> To catch something like this it might be an idea to add a very simple test target that just instantiates an "Amiga" object.

Agreed. I've adapted the console app for this purpose. On Windows, the `getopt` stuff is commented out and a default parameter used. With this change, the app compiles and links on Windows, too. 

Besides fixing the weird `dasmRead` issue, I only had to add `ws2_32` as target link library for MSVC and MinGW builds.

**Comment by mras0** [comment]
> Agreed. I've adapted the console app for this purpose. On Windows, the `getopt` stuff is commented out and a default parameter used. With this change, the app compiles and links on Windows, too.
> 
> Besides fixing the weird `dasmRead` issue, I only had to add `ws2_32` as target link library for MSVC and MinGW builds.

Great, the issue is resolved, and in the future you'll be pestered by MSVC first :)

Even if MSVC is right (it usually isn't when GCC disagrees), it's at best some esoteric C++ corner case that's not worth fuzzing about.
